ext/app/electron/config.ts

```diff
@@ -0,0 +1,206 @@
+import * as path from 'path'
+import * as fse from 'fs-extra'
+import * as electron from 'electron';
+import * as ini from 'ini'
+import * as os from 'os'
+import * as log from 'app/server/lib/log';
+import * as packageJson from 'desktop.package.json';
+import {commonUrls} from 'app/common/gristUrls';
+
+const CONFIG_DIR = path.join(electron.app.getPath("appData"), packageJson.name)
+
+const APPDATA_DIR = (process.platform == "win32") ? electron.app.getPath("userData"):
+  path.join(electron.app.getPath("home"), ".local", "share", packageJson.name)
+
+// Electron's app.getPath("userData") uses productName instead of name
+// but productName should be "full capitalized name", not ideal for naming our config directory.
+const DEFAULT_CONFIG_FILE = path.join(CONFIG_DIR, "config.ini")
+const NO_VALIDATION = () => true
+
+/**
+ * Suggest a value for an environment variable. If the variable is already set, do nothing. Otherwise set the value.
+ * @param name The name of the environment variable.
+ * @param value The value to suggest.
+ */
+function suggestEnv(name: string, value: string): void {
+  if (process.env[name] === undefined) {
+    process.env[name] = value;
+  }
+}
+
+type IConfig = {[key: string]: any}
+
+class Config implements IConfig {
+
+  [key: string]: any
+
+  constructor(c: IConfig) {
+    Object.assign(this, c)
+  }
+
+  /**
+   * Apply a configuration item by setting the corresponding environemt variable..
+   * The envvar, if already specified, has higher precedence over the config file.
+   *
+   * @param confKey The corresponding key in the config file.
+   * @param envKey The corresponding environment variable name.
+   * @param validator A predicate returning true if and only if the value to be used is valid. 
+   * @param defaultValue A default value to use when neither the environment variable nor the config file specifies a value, or when validation fails.
+   */
+  public apply(
+    confKey: string,
+    envKey: string,
+    validator: (value: string) => boolean,
+    defaultValue: string,
+  ): void {
+    let confValue: any = this
```

Comment:
Don't love this, would kind of prefer having a `this.confValue` rather than stretching to manipulate this